source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -110,11 +117,25 @@ absl::Status ProcessorState::processHeaderMutation(const CommonResponse& common_
 
 ProcessorState::CallbackState
 ProcessorState::getCallbackStateAfterHeaderResp(const CommonResponse& common_response) const {
-  if (bodyMode() == ProcessingMode::STREAMED &&
-      filter_.config().sendBodyWithoutWaitingForHeaderResponse() && !chunk_queue_.empty() &&
-      (common_response.status() != CommonResponse::CONTINUE_AND_REPLACE)) {
+  if (common_response.status() == CommonResponse::CONTINUE_AND_REPLACE) {
+    return ProcessorState::CallbackState::Idle;
+  }
+
+  if ((bodyMode() == ProcessingMode::STREAMED &&
+       filter_.config().sendBodyWithoutWaitingForHeaderResponse()) &&
+      !chunk_queue_.empty()) {
     return ProcessorState::CallbackState::StreamedBodyCallback;
   }
+
+  if (bodyMode() == ProcessingMode::FULL_DUPLEX_STREAMED) {
+    if (bodyReceived()) {
+      return ProcessorState::CallbackState::StreamedBodyCallback;
+    }
+    if (trailers_available_) {
+      return ProcessorState::CallbackState::TrailersCallback;
+    }
+  }
+
   return ProcessorState::CallbackState::Idle;
 }
 
```

Comment:
handleHeadersResponse is too big to comprehend. It wil be hard to know whether the change you made might have the desired effect, and no undesired ones.

WDYT of breaking this one up also?

Reply:
For test coverage, we should be good. These are the test cases we added for this new body mode:

Integration tests:

1) Server buffer headers, whole body, before sending response.
2) Server buffer headers, whole body, and trailers, before sending response.
3) Server buffer headers, and certain amount of body, then send send body response without wait of the end of it. At same time new body is coming in, and server continue do this kind of buffer-processing-response for a while. Then eventually trailers come in. Then server sends last chunk body response, and trailer response.

Unit tests:

1) Client sends header and body. Server send header response once receive header request, i.e, not waiting for body.
3) Client sends header and trailer, no body. Server sends header response after receiving trailer.
3) For one HTTP stream , server do MxN processing for some chunks, then do 1x1(i.e, send one response for one request immediately) for some chunks, then do MxN again.
4) A couple of server misbehaving test cases
5) A couple of Envoy misconfiguration test cases.

These tests are trying to cover different scenarios, like client requests may or may not have body, may or may not have trailer. Server sends header response may or may not wait for body, may or may not wait for trailers, et.al.

Reply:
Are we hitting all the lines in the coverage report?

Reply:
In terms of test coverage for ext_proc code in a whole, I recall it meets the Envoy criteria, like >96.3%. And also the ext_proc fuzzer coverage is > 70%

Reply:
That's good, but I'd still like you to look at the line coverage and see if we are hitting all your new code.

Reply:
Yes, based on this LCOV report, the new code are tested:
https://storage.googleapis.com/envoy-pr/8526639/coverage/source/extensions/filters/http/ext_proc/index.html